x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/ValuesSourceReaderOperator.java

```diff
@@ -123,44 +131,209 @@ public int count() {
             public int get(int i) {
                 return docs.getInt(i);
             }
-        }));
+        };
+        StoredFieldsSpec storedFieldsSpec = StoredFieldsSpec.NO_REQUIREMENTS;
+        List<RowStrideReaderWork> rowStrideReaders = new ArrayList<>(fields.size());
+        try {
+            for (int b = 0; b < fields.size(); b++) {
+                FieldWork field = fields.get(b);
+                BlockLoader.ColumnAtATimeReader columnAtATime = field.columnAtATime.reader(shard, segment, firstDoc);
+                if (columnAtATime != null) {
+                    blocks[b] = (Block) columnAtATime.read(blockFactory, loaderDocs);
+                } else {
+                    BlockLoader.RowStrideReader rowStride = field.rowStride.reader(shard, segment, firstDoc);
+                    rowStrideReaders.add(
+                        new RowStrideReaderWork(
+                            rowStride,
+                            (Block.Builder) field.info.blockLoaders.get(shard).builder(blockFactory, docs.getPositionCount()),
+                            b
+                        )
+                    );
+                    storedFieldsSpec = storedFieldsSpec.merge(field.info.blockLoaders.get(shard).rowStrideStoredFieldSpec());
+                }
+            }
+
+            if (rowStrideReaders.isEmpty()) {
+                return;
+            }
+            BlockLoaderStoredFieldsFromLeafLoader storedFields = null;
+            if (false == storedFieldsSpec.equals(StoredFieldsSpec.NO_REQUIREMENTS)) {
```

Comment:
I had sort of mimicked the behavior below where it should be possible to have row stride specs that don't need stored fields. But your right - this could just be an assert I think.

I agree with you about switching `null` for the empty stored fields loader. I'll do that in the block below and make this one an assert.